File: cpp/acroform/form_fields/form_fields.cpp
// form_fields -- C++ port of examples\Vb6\acroform\form_fields\form_fields.bas
#include <lumaspdf.h>
#include <cstdio>
#include <string>

static SI32 PDF_CALL PDFError(void* Data, SI32 ErrCode, const char* ErrMessage, SI32 ErrType){
    return 0;
}

static std::string exeDir(const char* a0){ std::string s(a0); auto p=s.find_last_of("\\/"); return p==std::string::npos?".":s.substr(0,p); }

int main(int argc, char** argv){
    SI32 f, r;
    double y;

    PPDF pdf = pdfNewPDF();
    pdfSetOnErrorProc(pdf, nullptr, PDFError);
    pdfCreateNewPDFA(pdf, "");

    pdfSetPageCoords(pdf, pcTopDown);

    pdfAppend(pdf);
    y = 50.0;
    pdfSetFontW(pdf, (LWCHAR*)LUMAS_TEXT("Helvetica"), fsRegular, 10.0, 0, cp1252);
    pdfWriteTextW(pdf, 50.0, y, (LWCHAR*)LUMAS_TEXT("Text fields:"));

    y = y + 15.0;
    f = pdfCreateTextField(pdf, "Text1", -1, 0, 0, 50.0, y, 200.0, 20.0);
    pdfSetTextFieldValueW(pdf, f, (LWCHAR*)LUMAS_TEXT(""), (LWCHAR*)LUMAS_TEXT("Single line text..."), taLeft);

    y = y + 30.0;
    f = pdfCreateTextField(pdf, "Text2", -1, 1, 0, 50.0, y, 200.0, 50.0);
    pdfSetTextFieldValueW(pdf, f, (LWCHAR*)LUMAS_TEXT(""), (LWCHAR*)LUMAS_TEXT("This field accepts multi-line text. The maximum text length can be restricted if necessary."), taLeft);

    y = y + 60.0;
    pdfWriteTextW(pdf, 50.0, y, (LWCHAR*)LUMAS_TEXT("A password field:"));
    y = y + 15.0;
    f = pdfCreateTextField(pdf, "Text3", -1, 0, 0, 50.0, y, 200.0, 20.0);
    pdfSetFieldFlags(pdf, f, ffPassword, 0);
    pdfSetTextFieldValueW(pdf, f, (LWCHAR*)LUMAS_TEXT(""), (LWCHAR*)LUMAS_TEXT("**********"), taLeft);

    y = y + 30.0;
    pdfWriteTextW(pdf, 50.0, y, (LWCHAR*)LUMAS_TEXT("A fixed length field separated into combs:"));
    y = y + 15.0;
    f = pdfCreateTextField(pdf, "Text4", -1, 0, 10, 50.0, y, 200.0, 20.0);
    pdfSetFieldFlags(pdf, f, ffComb, 0);

    y = 50.0;
    pdfWriteTextW(pdf, 350.0, y, (LWCHAR*)LUMAS_TEXT("Choice fields:"));
    y = y + 15.0;
    f = pdfCreateComboBox(pdf, "Combo1", 1, -1, 350.0, y, 200.0, 20.0);
    pdfAddValToChoiceFieldW(pdf, f, "", (LWCHAR*)LUMAS_TEXT(" Select a value..."), 1);
    pdfAddValToChoiceFieldW(pdf, f, "Apple", (LWCHAR*)LUMAS_TEXT("Apple"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Banana", (LWCHAR*)LUMAS_TEXT("Banana"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Pear", (LWCHAR*)LUMAS_TEXT("Pear"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Grape", (LWCHAR*)LUMAS_TEXT("Grape"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Orange", (LWCHAR*)LUMAS_TEXT("Orange"), 0);

    y = y + 30.0;
    f = pdfCreateListBox(pdf, "List", 1, -1, 350.0, y, 200.0, 50.0);
    pdfAddValToChoiceFieldW(pdf, f, "Apple", (LWCHAR*)LUMAS_TEXT("Apple"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Banana", (LWCHAR*)LUMAS_TEXT("Banana"), 1);
    pdfAddValToChoiceFieldW(pdf, f, "Pear", (LWCHAR*)LUMAS_TEXT("Pear"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Grape", (LWCHAR*)LUMAS_TEXT("Grape"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Orange", (LWCHAR*)LUMAS_TEXT("Orange"), 0);

    y = y + 60.0;
    pdfWriteTextW(pdf, 350.0, y, (LWCHAR*)LUMAS_TEXT("Editable combo box:"));
    y = y + 15.0;
    f = pdfCreateComboBox(pdf, "Combo2", 1, -1, 350.0, y, 200.0, 20.0);
    pdfAddValToChoiceFieldW(pdf, f, "Apple", (LWCHAR*)LUMAS_TEXT("Apple"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Banana", (LWCHAR*)LUMAS_TEXT("Banana"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Pear", (LWCHAR*)LUMAS_TEXT("Pear"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Grape", (LWCHAR*)LUMAS_TEXT("Grape"), 0);
    pdfAddValToChoiceFieldW(pdf, f, "Orange", (LWCHAR*)LUMAS_TEXT("Orange"), 0);
    pdfSetFieldFlags(pdf, f, ffEdit, 0);
    pdfSetFieldExpValueW(pdf, f, 1000, (LWCHAR*)LUMAS_TEXT("Select or enter a value..."), "", 1);

    y = y + 30.0;
    pdfWriteTextW(pdf, 350.0, y, (LWCHAR*)LUMAS_TEXT("Check boxes / Radio buttons:"));

    y = y + 15.0;
    pdfChangeFontSize(pdf, 1.0);
    pdfCreateCheckBox(pdf, "N1", "C1", 1, -1, 350.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "N2", "C2", 1, -1, 380.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "N3", "C1", 1, -1, 410.0, y, 20.0, 20.0);

    pdfCreateCheckBox(pdf, "G1", "C1", 0, -1, 450.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 480.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "G1", "C1", 1, -1, 510.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 540.0, y, 20.0, 20.0);

    y = y + 30.0;
    pdfChangeFontSize(pdf, 15.0);
    pdfSetCheckBoxChar(pdf, ccCircle);
    r = pdfCreateRadioButton(pdf, "Radio1", "R1", 1, -1, 350.0, y, 20.0, 20.0);
    pdfSetCheckBoxDefState(pdf, r, 0);
    pdfCreateCheckBox(pdf, "", "R2", 0, r, 380.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "", "R3", 0, r, 410.0, y, 20.0, 20.0);

    r = pdfCreateRadioButton(pdf, "Radio2", "R1", 1, -1, 450.0, y, 20.0, 20.0);
    pdfSetFieldFlags(pdf, r, ffRadioIsUnion, 0);
    pdfCreateCheckBox(pdf, "", "R2", 0, r, 480.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "", "R1", 1, r, 510.0, y, 20.0, 20.0);
    pdfCreateCheckBox(pdf, "", "R2", 0, r, 540.0, y, 20.0, 20.0);
    pdfEndPage(pdf);

    if(pdfHaveOpenDoc(pdf) != 0){
        std::string outFile = exeDir(argv[0]) + "/out.pdf";
        if(pdfOpenOutputFileA(pdf, outFile.c_str()) == 0){
            pdfDeletePDF(pdf);
            return 0;
        }
        if(pdfCloseFile(pdf) != 0)
            printf("PDF file \"%s\" successfully created!\n", outFile.c_str());
    }

    pdfDeletePDF(pdf);
    return 0;
}

File: csharp/acroform/form_fields/form_fields.cs
//  form_fields -- C# port of examples\Vb6\acroform\form_fields\form_fields.bas
//  Text fields (single/multi-line/password/comb), combo boxes, list boxes,
//  editable combo, check boxes and radio buttons.
using System;
using System.IO;
using LumasPdfSdk;

class FormFields
{
    static readonly TErrorProc _err = PDFError;
    static int PDFError(IntPtr Data, int ErrCode, string ErrMessage, int ErrType) { return 0; }

    static void Main()
    {
        double y;
        IntPtr pdf = LumasPdf.pdfNewPDF();

        LumasPdf.pdfSetOnErrorProc(pdf, IntPtr.Zero, _err);
        LumasPdf.pdfCreateNewPDFW(pdf, "");

        LumasPdf.pdfSetPageCoords(pdf, (int)TPageCoord.pcTopDown);

        LumasPdf.pdfAppend(pdf);
        y = 50.0;
        LumasPdf.pdfSetFontW(pdf, "Helvetica", LumasPdfConsts.fsRegular, 10.0, false, TCodepage.cp1252);
        LumasPdf.pdfWriteTextW(pdf, 50, y, "Text fields:");

        y = y + 15.0;
        int f = LumasPdf.pdfCreateTextField(pdf, "Text1", -1, 0, 0, 50, y, 200, 20);
        LumasPdf.pdfSetTextFieldValueW(pdf, (uint)f, "", "Single line text...", LumasPdfConsts.taLeft);

        y = y + 30.0;
        f = LumasPdf.pdfCreateTextField(pdf, "Text2", -1, 1, 0, 50, y, 200, 50);
        LumasPdf.pdfSetTextFieldValueW(pdf, (uint)f, "", "This field accepts multi-line text. The maximum text length can be restricted if necessary.", LumasPdfConsts.taLeft);

        y = y + 60.0;
        LumasPdf.pdfWriteTextW(pdf, 50, y, "A password field:");
        y = y + 15.0;
        f = LumasPdf.pdfCreateTextField(pdf, "Text3", -1, 0, 0, 50, y, 200, 20);
        LumasPdf.pdfSetFieldFlags(pdf, (uint)f, (int)LumasPdfConsts.ffPassword, false);
        LumasPdf.pdfSetTextFieldValueW(pdf, (uint)f, "", "**********", LumasPdfConsts.taLeft);

        y = y + 30.0;
        LumasPdf.pdfWriteTextW(pdf, 50, y, "A fixed length field separated into combs:");
        y = y + 15.0;
        f = LumasPdf.pdfCreateTextField(pdf, "Text4", -1, 0, 10, 50, y, 200, 20);
        LumasPdf.pdfSetFieldFlags(pdf, (uint)f, (int)LumasPdfConsts.ffComb, false);

        y = 50.0;
        LumasPdf.pdfWriteTextW(pdf, 350, y, "Choice fields:");
        y = y + 15.0;
        f = LumasPdf.pdfCreateComboBox(pdf, "Combo1", 1, -1, 350, y, 200, 20);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "", " Select a value...", 1);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Apple", "Apple", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Banana", "Banana", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Pear", "Pear", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Grape", "Grape", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Orange", "Orange", 0);

        y = y + 30.0;
        f = LumasPdf.pdfCreateListBox(pdf, "List", true, -1, 350, y, 200, 50);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Apple", "Apple", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Banana", "Banana", 1);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Pear", "Pear", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Grape", "Grape", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Orange", "Orange", 0);

        y = y + 60.0;
        LumasPdf.pdfWriteTextW(pdf, 350, y, "Editable combo box:");
        y = y + 15.0;
        f = LumasPdf.pdfCreateComboBox(pdf, "Combo2", 1, -1, 350, y, 200, 20);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Apple", "Apple", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Banana", "Banana", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Pear", "Pear", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Grape", "Grape", 0);
        LumasPdf.pdfAddValToChoiceFieldW(pdf, (uint)f, "Orange", "Orange", 0);
        LumasPdf.pdfSetFieldFlags(pdf, (uint)f, (int)LumasPdfConsts.ffEdit, false);
        LumasPdf.pdfSetFieldExpValueW(pdf, (uint)f, 1000, "Select or enter a value...", "", true);

        y = y + 30.0;
        LumasPdf.pdfWriteTextW(pdf, 350, y, "Check boxes / Radio buttons:");

        y = y + 15.0;
        LumasPdf.pdfChangeFontSize(pdf, 1.0);
        LumasPdf.pdfCreateCheckBox(pdf, "N1", "C1", 1, -1, 350, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "N2", "C2", 1, -1, 380, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "N3", "C1", 1, -1, 410, y, 20, 20);

        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C1", 0, -1, 450, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 480, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C1", 1, -1, 510, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "G1", "C2", 0, -1, 540, y, 20, 20);

        y = y + 30.0;
        LumasPdf.pdfChangeFontSize(pdf, 15.0);
        LumasPdf.pdfSetCheckBoxChar(pdf, (int)TCheckBoxChar.ccCircle);
        int r = LumasPdf.pdfCreateRadioButton(pdf, "Radio1", "R1", 1, -1, 350, y, 20, 20);
        LumasPdf.pdfSetCheckBoxDefState(pdf, (uint)r, false);
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 380, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "", "R3", 0, r, 410, y, 20, 20);

        r = LumasPdf.pdfCreateRadioButton(pdf, "Radio2", "R1", 1, -1, 450, y, 20, 20);
        LumasPdf.pdfSetFieldFlags(pdf, (uint)r, (int)LumasPdfConsts.ffRadioIsUnion, false);
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 480, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "", "R1", 1, r, 510, y, 20, 20);
        LumasPdf.pdfCreateCheckBox(pdf, "", "R2", 0, r, 540, y, 20, 20);
        LumasPdf.pdfEndPage(pdf);

        if (LumasPdf.pdfHaveOpenDoc(pdf))
        {
            string outFile = Path.Combine(AppDomain.CurrentDomain.BaseDirectory, "out.pdf");
            if (!LumasPdf.pdfOpenOutputFileW(pdf, outFile)) { LumasPdf.pdfDeletePDF(pdf); return; }
            if (LumasPdf.pdfCloseFile(pdf))
                Console.WriteLine("PDF file \"" + outFile + "\" successfully created!");
        }

        LumasPdf.pdfDeletePDF(pdf);
    }
}
